clang/lib/Sema/SemaTemplateDeductionGuide.cpp

```diff
@@ -931,12 +935,73 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
   return {Template, AliasRhsTemplateArgs};
 }
 
+struct InheritedConstructorDeductionInfo {
+  // Class template for which we are declaring deduction guides
+  // This is `C` in the standard wording
+  TemplateDecl *DerivedClassTemplate;
+
+  // `template<typename> CC` in the standard wording
+  // This is the type of template that is substituted in the deduction guide
+  // return type `CC<R>`
+  TypeSourceInfo *CCType;
+};
+
+// Build the type for a deduction guide generated from an inherited constructor
+// C++23 [over.match.class.deduct]p1.10:
+// ... the set contains the guides of A with the return type R
+// of each guide replaced with `typename CC<R>::type` ...
+TypeSourceInfo *buildInheritedConstructorDeductionGuideType(
+    Sema &SemaRef, const InheritedConstructorDeductionInfo &Info,
+    TypeSourceInfo *SourceGuideTSI) {
+  auto &Context = SemaRef.Context;
+  const auto *FPT = SourceGuideTSI->getType()->getAs<FunctionProtoType>();
+  assert(FPT && "Source Guide type should be a FunctionProtoType");
+
+  // This substitution can fail in cases where the source return type
+  // is not dependent and the derived class is not deducible
+  Sema::SFINAETrap Trap(SemaRef);
```

Comment:
This is not clear to me, why would the substitution fail if the source is not dependent? Would there instead be simply no substitution? Also why is the derived class not being deducible relevant here, since we are not performing any deductions?

Reply:
This code is substituting the return type `R` of the base deduction guide into the CC template type `typename CC<R>::type` to form the return type for the inherited deduction guide. When `R` is dependent, the resulting type is also dependent, but if `R` is not dependent this will cause the CC template to be instantiated here.

The motivating test case for when this is problematic is:
```c++
template <typename T>
struct NonTemplateDGuideBase {
    NonTemplateDGuideBase(T);
};

NonTemplateDGuideBase(int) -> NonTemplateDGuideBase<int>;
NonTemplateDGuideBase(const char *) -> NonTemplateDGuideBase<const char *>;

template <typename T>
concept NoPointers = !requires(T t) { *t; };

template <NoPointers T>
struct NonTemplateDGuideDerived : public NonTemplateDGuideBase<T> {
    using NonTemplateDGuideBase<T>::NonTemplateDGuideBase;
};

NonTemplateDGuideDerived ntdg_char("");
```

Since constraints are involved, substituting a non-dependent `R` into the CC type causes constraints to be evaluated, and if the derived class is not deducible given base type `R` (e.g. when the constraints evaluate to false), this substitution will fail. This is not an issue if `R` is dependent as the constraints will evaluate later during argument deduction, or if the derived class is deducible given `R` as `CC<R>` will be a valid substitution (e.g. in the above example, the `int` deduction guide does not cause issues).

Reply:
I see. But why then are we suppressing the diagnostics in this case?
If there is a substitution failure otherwise, is that currently being diagnosed somewhere else?

Reply:
If we don't suppress diagnostics here, then these substitution failure errors would appear at the time that implicit deduction guides are declared (regardless of whether or not the failing deduction guide was actually used).

Right now any deduction guide that fails substitution here will be skipped, and not inherited by the base. Unfortunately this means that we don't get diagnostics for them, but there isn't a great way to propagate the error through to the point where deduction of the derived class fails. We don't have the return type to create a valid deduction guide AST node, like we do in other cases where we know deduction will fail up front (e.g. with default parameters not appearing in the base), so some other mechanism is needed to preserve this information until deduction of the derived class has failed and the candidate notes are being printed.

Reply:
I see, that is a bummer. This is worth pointing out in a FIXME.
Are you planning on working on a fix for that?

Reply:
I have added a FIXME, I can work on a fix for it in a followup but it will likely need more invasive changes to the AST in order to preserve the necessary information.